### ISSUE TITLE
Prevent Vue Router from accessing the history object when using hash mode

### ISSUE THREAD

**Issue by PierBover** [vuejs/vue-router#2304]
### What problem does this feature solve?
The feature request is simple: Vue Router should not access `history` at all when it starts in hash mode.

**But Why?**

The problem is that Chrome web apps (as in apps installed from the Chrome Web Store to Chrome OS devices) do not have access to the same browser API a regular web app has. In this particular case the problem is that the `history` object is not present.

This was not really a problem before, `history` was `undefined` and it didn't really matter. The current behavior is different. When an app tries to access the `history` object it completely crashes. A polyfill cannot be implemented either because of this.

It would seem the solution to this problem should be simply to stick to hash mode, but Vue Router still tries to access `history` even in hash mode. I imagine this is done to check whether the history API is present, but since hash mode is already specified, those checks are unnecessary.

A couple of months ago I hacked my way through this by modifying Vue Router. If there is interest I could implement this feature myself.

Even if this feature is not implemented the Vue team should at least be aware of this limitation.

### What does the proposed API look like?
No changes in the API would be necessary.

**Comment by LinusBorg** [comment]
> When an app tries to access the history object it completely crashes.

I would assume that's a bug with that platform,  right?

> It would seem the solution to this problem should be simply to stick to hash mode, but Vue Router still tries to access `history` even in hash mode. I imagine this is done to check whether the history API is present, but since hash mode is already specified, those checks are unnecessary.

Not really. One might use hash mode, but want to use `scrollBehaviour` - which only works if `pushState` is supported. Checking that requires accessing the `window.history` object.

See here:

https://github.com/vuejs/vue-router/blob/dev/src/history/hash.js#L25
https://github.com/vuejs/vue-router/blob/dev/src/util/push-state.js#L18

Removing that would result in a breaking change that makes `scrollBehaviour` unavailable in `hash` mode

So I don't think this is feasable

**Comment by PierBover** [comment]
> I would assume that's a bug with that platform, right?

I doubt it, it seems to me it's Chrome OS preventing access to certain APIs. A couple of years back it was a silent error. Now the app simply cannot continue execution. It's not only related to `window.history`, there are a number of other APIs such as `window.confirm` that aren't available either.

> Not really. One might use hash mode, but want to use scrollBehaviour - which only works if pushState is supported. Checking that requires accessing the window.history object.

Indeed.

I forked [Vue Router](https://github.com/PierBover/vue-router) to be able to use it in Chrome OS, and I solved the scroll problem with:

```
watch: {
	$route (to, from) {
		this.$refs['some-div'].scrollTop = 0;
	}
}
```

Anyway, I completely understand that offering limited support to Chrome OS apps might not be in the interest of the Vue team though.
